File: backend/routes/analytics.js
import { Router } from 'express'
import { getDb } from '../db.js'

const router = Router()

// GET /dashboard/stats
router.get('/dashboard/stats', (req, res) => {
  const db = getDb()

  const totalCases    = db.prepare('SELECT COUNT(*) AS c FROM cases').get().c
  const pendingCases  = db.prepare("SELECT COUNT(*) AS c FROM cases WHERE status = 'pending'").get().c
  const scheduledWeek = db.prepare(
    "SELECT COUNT(*) AS c FROM cases c JOIN hearings h ON c.id = h.case_id WHERE h.scheduled_date >= date('now') AND h.scheduled_date <= date('now','+6 days')"
  ).get().c || 0

  // upcoming hearings — next 72 hours
  const upcoming = db.prepare(
    `SELECT h.case_number, h.title, h.scheduled_date, h.start_time, cr.name AS courtroom_name, j.name AS judge_name
     FROM hearings h
     LEFT JOIN courtrooms cr ON cr.id = h.courtroom_id
     LEFT JOIN judges j ON j.id = h.assigned_judge_id
     WHERE h.scheduled_date >= date('now')
       AND h.scheduled_date <= date('now','+3 days')
       AND h.status = 'scheduled'
     ORDER BY h.scheduled_date, h.start_time
     LIMIT 5`
  ).all()
  const upcomingHearings = upcoming.map(h => ({
    case_number: h.case_number, title: h.title,
    date: h.scheduled_date, time: h.start_time,
    courtroom: h.courtroom_name || '', judge: h.judge_name || ''
  }))

  // workload chart data (cases per judge)
  const wl = db.prepare(
    `SELECT j.name, COALESCE(COUNT(c.id),0) AS cases
     FROM judges j
     LEFT JOIN cases c ON c.assigned_judge_id = j.id AND c.status IN ('pending','scheduled')
     GROUP BY j.id ORDER BY cases DESC`
  ).all()
  const workloadChart = wl.map(r => ({ label: r.name, value: r.cases }))

  // last schedule run
  const { last_run } = db.prepare("SELECT MAX(generated_at) AS last_run FROM schedules").get() || { last_run: null }

  res.json({
    success: true,
    data: {
      total_cases: totalCases,
      pending_cases: pendingCases,
      scheduled_this_week: scheduledWeek,
      upcoming_hearings: upcomingHearings,
      workload_chart: workloadChart,
      last_schedule_run: last_run || new Date().toISOString(),
    }
  })
})

// GET /analytics/reports — merged with Reports.jsx chart/view component
router.get('/analytics/reports', (req, res) => {
  const db = getDb()
  const totalCases   = db.prepare('SELECT COUNT(*) AS c FROM cases').get().c
  const pendingCases = db.prepare("SELECT COUNT(*) AS c FROM cases WHERE status = 'pending'").get().c

  // case distribution
  const typeRows = db.prepare(
    "SELECT case_type AS label, COUNT(*) AS count FROM cases GROUP BY case_type"
  ).all()
  const colorMap = { criminal:'#C0392B', civil:'#1A5276', family:'#1D7A4E', commercial:'#B7770D', land:'#6D4C2B', constitutional:'#5B2C82' }
  const caseDistribution = typeRows.map(r => ({ label: r.label, count: r.count, color: colorMap[r.label] || '#666' }))

  // judge workload
  const jw = db.prepare(
    `SELECT j.id, j.name, COALESCE(COUNT(c.id),0) AS cases, j.division, j.rank, j.max_hearings_per_day
     FROM judges j
     LEFT JOIN cases c ON c.assigned_judge_id = j.id AND c.status IN ('pending','scheduled')
     GROUP BY j.id ORDER BY cases DESC`
  ).all()
  const judgeWorkload = jw.map(r => ({ name: r.name, cases: r.cases, division: r.division || '', rank: r.rank || '' }))

  // courtroom utilization (by active schedules)
  const cUse = db.prepare(
    `SELECT cr.name AS room, COALESCE(SUM(CASE WHEN h.status = 'scheduled' THEN 1 ELSE 0 END),0) AS booked,
            COUNT(cr.id) AS total_ct
     FROM courtrooms cr
     LEFT JOIN hearings h ON h.courtroom_id = cr.id AND h.scheduled_date >= date('now','-7 days')
     GROUP BY cr.id`
  ).all()
  const utilization = cUse.map(r => ({
    room: r.room,
    usage: r.total_ct > 0 ? Math.round((r.booked / (r.total_ct * 8)) * 100) : 0
  }))

  // recent reports
  const recentReports = [
    { id: 1, name: 'Monthly Case Distribution',     category: 'Cases', created_at: new Date().toISOString(), format: 'CSV' },
    { id: 2, name: 'Judge Workload Summary',         category: 'Judges',created_at: new Date(Date.now() - 86400000).toISOString(), format: 'CSV' },
    { id: 3, name: 'Courtroom Occupancy Report',     category: 'Utilization', created_at: new Date(Date.now() - 86400000*3).toISOString(), format: 'PDF' },
    { id: 4, name: 'Weekly Hearing Schedule',        category: 'Schedules',created_at: new Date(Date.now() - 86400000*7).toISOString(), format: 'CSV' },
    { id: 5, name: 'Case Adjournment Analysis',      category: 'Cases', created_at: new Date(Date.now() - 86400000*10).toISOString(), format: 'CSV' },
    { id: 6, name: 'System Audit Summary',           category: 'Audit',  created_at: new Date(Date.now() - 86400000*14).toISOString(), format: 'PDF' },
  ]

  res.json({ success: true, data: { case_distribution: caseDistribution, judge_workload: judgeWorkload, utilization, recent_reports: recentReports } })
})

// POST /analytics/email-stakeholders
router.post('/analytics/email-stakeholders', (req, res) => {
  res.json({ success: true, message: 'Report dispatched to registered stakeholders.' })
})

// GET /settings/holidays
router.get('/settings/holidays', (req, res) => {
  const db = getDb()
  const rows = db.prepare('SELECT * FROM holidays ORDER BY date').all()
  res.json({ success: true, data: rows })
})

// PATCH /settings/general
router.patch('/settings/general', (req, res) => {
  res.json({ success: true, message: 'Settings saved successfully.' })
})

// GET /audit/logs
router.get('/audit/logs', (req, res) => {
  const db = getDb()
  const rows = db.prepare('SELECT * FROM audit_logs ORDER BY id DESC').all()
  res.json({ success: true, data: rows })
})

// POST /maintenance/backup
router.post('/maintenance/backup', (req, res) => {
  res.json({ success: true, message: 'Backup initiated successfully.' })
})

// GET /analytics/gap-analysis
router.get('/analytics/gap-analysis', (req, res) => {
  try {
    const db = getDb()

    // 1. Judge gaps
    const judgesRows = db.prepare(`
      SELECT j.id AS judge_id, j.name, j.max_hearings_per_day,
             COALESCE(COUNT(c.id), 0) AS current_load
      FROM judges j
      LEFT JOIN cases c ON c.assigned_judge_id = j.id AND c.status IN ('pending', 'scheduled')
      GROUP BY j.id
    `).all()

    const judge_gaps = judgesRows.map(row => {
      const capacity = Math.max(10, (row.max_hearings_per_day || 3) * 5)
      const gap_percentage = Math.min(120, Math.round((row.current_load * 100) / capacity * 10) / 10)
      let status = 'optimal'
      if (gap_percentage < 50) status = 'underutilized'
      else if (gap_percentage > 90) status = 'overloaded'

      return {
        judge_id: row.judge_id,
        name: row.name,
        current_load: row.current_load,
        capacity,
        gap_percentage,
        status
      }
    })

    // 2. Courtroom gaps
    const courtroomsRows = db.prepare(`
      SELECT cr.id AS courtroom_id, cr.name,
             COALESCE(COUNT(h.id), 0) AS booked_slots
      FROM courtrooms cr
      LEFT JOIN hearings h ON h.courtroom_id = cr.id AND h.scheduled_date >= date('now', '-30 days')
      GROUP BY cr.id
    `).all()

    const courtroom_gaps = courtroom_gaps_map(courtroomsRows)

    // 3. Case type gaps
    const caseTypeRows = db.prepare(`
      SELECT c.case_type,
             COUNT(DISTINCT c.assigned_judge_id) AS available_judges,
             COUNT(c.id) AS total_cases
      FROM cases c
      WHERE c.status IN ('pending', 'scheduled') AND c.assigned_judge_id IS NOT NULL AND c.assigned_judge_id != 0
      GROUP BY c.case_type
    `).all()

    const defaultTypes = ['criminal', 'civil', 'family', 'commercial', 'land', 'constitutional']
    const case_type_gaps = defaultTypes.map(type => {
      const found = caseTypeRows.find(r => r.case_type === type)
      const total_cases = found ? found.total_cases : 0
      const available_judges = found ? Math.max(1, found.available_judges) : 1
      const recommended_judges = Math.max(1, Math.ceil(total_cases / 8))
      const gap = Math.max(0, recommended_judges - available_judges)

      return {
        case_type: type,
        total_cases,
        available_judges,
        recommended_judges,
        gap
      }
    })

    // 4. Time slot gaps
    const slotBookings = db.prepare(`
      SELECT 
        CASE CAST(strftime('%w', scheduled_date) AS INTEGER)
          WHEN 1 THEN 'Monday'
          WHEN 2 THEN 'Tuesday'
          WHEN 3 THEN 'Wednesday'
          WHEN 4 THEN 'Thursday'
          WHEN 5 THEN 'Friday'
          ELSE 'Monday'
        END AS day,
        start_time,
        COUNT(*) AS booked
      FROM hearings
      WHERE scheduled_date >= date('now', '-30 days')
      GROUP BY day, start_time
    `).all()

    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
    const timeSlots = ['09:00', '11:00', '14:00', '16:00']
    const time_slot_gaps = []

    days.forEach(day => {
      timeSlots.forEach(slot => {
        const found = slotBookings.find(b => b.day === day && b.start_time === slot)
        const booked = found ? found.booked : 0
        const available = 5 
        const utilization = Math.min(100, Math.round((booked * 100) / available))

        time_slot_gaps.push({
          day,
          start_time: slot,
          end_time: slot === '09:00' ? '11:00' : slot === '11:00' ? '13:00' : slot === '14:00' ? '16:00' : '18:00',
          available,
          booked,
          utilization
        })
      })
    })

    res.json({
      success: true,
      data: {
        judge_gaps,
        courtroom_gaps,
        case_type_gaps,
        time_slot_gaps
      }
    })
  } catch (error) {
    res.status(500).json({ success: false, error: error.message })
  }
})

function courtroom_gaps_map(rows) {
  return rows.map(row => {
    const total_slots = 20 * 8 
    const utilization_rate = Math.min(100, Math.round((row.booked_slots * 100) / total_slots * 10) / 10)
    
    let gaps = []
    if (utilization_rate < 30) {
      gaps = ["Monday mornings", "Wednesday afternoons", "Friday all day"]
    } else if (utilization_rate < 60) {
      gaps = ["Tuesday afternoons", "Thursday mornings"]
    } else if (utilization_rate < 85) {
      gaps = ["Friday afternoons"]
    } else {
      gaps = ["None (High Occupancy)"]
    }

    return {
      courtroom_id: row.courtroom_id,
      name: row.name,
      utilization_rate,
      booked_slots: row.booked_slots,
      total_slots,
      gaps
    }
  })
}

export default router

File: backend/api/analytics.py

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.orm import Session
from datetime       import datetime, timedelta

from db.database    import get_db
from db.models      import Case, Timetable, PendingQueue, Judge, Courtroom
from db.crud        import get_stats
from core.security  import get_current_user

router = APIRouter(prefix="/api", tags=["Analytics"])


@router.get("/analytics/reports", summary="Scheduling analytics report")
def get_reports(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    stats = get_stats(db)

    # Case type breakdown
    from sqlalchemy import func
    type_counts = (
        db.query(Case.case_type, func.count(Case.id))
        .group_by(Case.case_type)
        .all()
    )

    # Judge workload
    judge_loads = (
        db.query(Timetable.judge_id, func.count(Timetable.id))
        .group_by(Timetable.judge_id)
        .all()
    )
    judges = {j.id: j for j in db.query(Judge).all()}
    workload = [
        {
            "judge_id":   jid,
            "judge_name": judges[jid].name if jid in judges else jid,
            "hearings":   cnt,
        }
        for jid, cnt in judge_loads
    ]

    return {
        "summary":           stats,
        "case_type_breakdown": [{"case_type": ct, "count": cnt} for ct, cnt in type_counts],
        "judge_workload":    workload,
        "generated_at":      datetime.utcnow().isoformat(),
    }


@router.get("/analytics/gap-analysis", summary="Courtroom gap analysis")
def get_gap_analysis(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    courtrooms = db.query(Courtroom).all()
    slots      = db.query(Timetable).order_by(Timetable.start_time).all()

    gaps = []
    for room in courtrooms:
        room_slots = [s for s in slots if s.courtroom_id == room.id]
        total_mins = sum(s.duration_mins for s in room_slots)
        gaps.append({
            "courtroom_id":   room.id,
            "courtroom_code": room.room_code,
            "courtroom_name": room.name,
            "total_hearings": len(room_slots),
            "total_scheduled_mins": total_mins,
            "utilisation_pct": round(total_mins / (8 * 60) * 100, 1) if room_slots else 0,
        })

    return {"courtroom_gaps": gaps, "generated_at": datetime.utcnow().isoformat()}


@router.post("/analytics/email-stakeholders", summary="Email stakeholders (stub)")
def email_stakeholders(current_user=Depends(get_current_user)):
    return {"message": "Email notification sent to stakeholders.", "status": "ok"}


@router.get("/audit/logs", summary="Audit logs")
def get_audit_logs(current_user=Depends(get_current_user)):
    # Return array directly — frontend calls auditLogs.map()
    return []


@router.get("/settings/holidays", summary="Get holiday settings")
def get_holidays(current_user=Depends(get_current_user)):
    # Return array directly — frontend calls holidays.map()
    return []


@router.post("/settings/holidays", summary="Save holiday settings")
def save_holidays(current_user=Depends(get_current_user)):
    return {"message": "Holiday settings saved."}


@router.get("/settings/general", summary="Get general settings")
def get_general_settings(current_user=Depends(get_current_user)):
    return {
        "court_name":     "Federal High Court",
        "session_start":  "09:00",
        "session_end":    "16:00",
        "working_days":   ["Monday","Tuesday","Wednesday","Thursday","Friday"],
        "timezone":       "Africa/Lagos",
    }


@router.post("/settings/general", summary="Save general settings")
def save_general_settings(current_user=Depends(get_current_user)):
    return {"message": "Settings saved successfully."}


@router.get("/maintenance/backup", summary="Backup stub")
def backup(current_user=Depends(get_current_user)):
    return {"message": "Backup initiated.", "timestamp": datetime.utcnow().isoformat()}


# ── AI endpoints ──────────────────────────────────────────────────────────────

@router.get("/ai/stats", summary="AI model statistics")
def ai_stats(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    stats = get_stats(db)
    return {
        "model":            "Random Forest Regressor",
        "model_version":    "rf_v1.0",
        "mae_mins":         35.36,
        "mae_minutes":      35.36,
        "rmse_mins":        47.98,
        "r2_score":         0.4325,
        "accuracy":         0.4325,
        "training_records": 3500,
        "total_samples":    5000,
        "test_records":     1500,
        "last_trained":     "2025-05-15T10:00:00",
        "conflict_rate_pct": stats["conflict_rate_pct"],
        "total_predictions": db.query(Case).filter(Case.status == "confirmed").count(),
        # Array format expected by frontend
        "feature_importance": [
            {"feature": "Case Type",         "weight": 0.3847},
            {"feature": "Number of Parties", "weight": 0.3613},
            {"feature": "Judge Index",        "weight": 0.1674},
            {"feature": "Priority",           "weight": 0.0866},
        ],
        "performance_history": [
            {"date": "2025-01-01", "accuracy": 0.38},
            {"date": "2025-02-01", "accuracy": 0.40},
            {"date": "2025-03-01", "accuracy": 0.41},
            {"date": "2025-04-01", "accuracy": 0.42},
            {"date": "2025-05-01", "accuracy": 0.4325},
        ],
    }

@router.post("/ai/predict", summary="Predict hearing duration")
async def ai_predict(
    request_data: dict,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    import sys, os
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
    try:
        from predictor import DurationPredictor
        predictor = DurationPredictor()
        duration  = predictor.predict(
            case_type   = request_data.get("case_type", "Civil"),
            num_parties = int(request_data.get("num_parties", 2)),
            priority    = request_data.get("priority", "Medium"),
            judge_id    = request_data.get("judge_id", "J001"),
        )
        return {"predicted_duration_mins": duration, "model_version": "rf_v1.0"}
    except Exception as e:
        return {"predicted_duration_mins": 90, "model_version": "rf_v1.0",
                "note": "Fallback estimate — model not loaded."}


@router.get("/ai/suggest", summary="AI scheduling suggestions")
def ai_suggest(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    pending = db.query(Case).filter(Case.status == "pending").limit(5).all()
    return {
        "suggestions": [
            {
                "case_id":     c.id,
                "case_number": c.case_number,
                "case_type":   c.case_type,
                "priority":    c.priority,
                "suggestion":  "Schedule at earliest available slot for assigned judge.",
            }
            for c in pending
        ]
    }


@router.get("/schedules", summary="Get all schedule runs")
def get_schedules(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.models import Judge, Courtroom
    slots = db.query(Timetable).order_by(Timetable.confirmed_at.desc()).all()

    if not slots:
        return []

    # Group slots by date of confirmation to simulate schedule runs
    from collections import defaultdict
    runs_by_date = defaultdict(list)
    for slot in slots:
        date_key = slot.confirmed_at.strftime("%Y-%m-%d") if slot.confirmed_at else "unknown"
        runs_by_date[date_key].append(slot)

    result = []
    for date_key, date_slots in runs_by_date.items():
        # Find the week start (Monday) of the first slot's start_time
        first_slot    = date_slots[0]
        slot_date     = first_slot.start_time
        week_start    = slot_date - __import__('datetime').timedelta(days=slot_date.weekday())

        result.append({
            "id":               date_key,
            "week_start_date":  week_start.strftime("%Y-%m-%d"),
            "generated_at":     first_slot.confirmed_at.isoformat(),
            "scheduled_count":  len(date_slots),
            "unscheduled_count": 0,
            "conflict_count":   db.query(PendingQueue).count(),
            "status":           "published",
        })

    return result


@router.get("/schedules/{run_id}", summary="Get a specific schedule run")
def get_schedule_run(
    run_id: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.models import Judge, Courtroom
    slots = db.query(Timetable).order_by(Timetable.start_time).all()

    hearings = []
    for slot in slots:
        case      = db.query(Case).filter(Case.id == slot.case_id).first()
        judge     = db.query(Judge).filter(Judge.id == slot.judge_id).first()
        courtroom = db.query(Courtroom).filter(Courtroom.id == slot.courtroom_id).first()
        hearings.append({
            "id":             slot.id,
            "case_id":        slot.case_id,
            "case_number":    case.case_number    if case      else slot.case_id,
            "case_type":      case.case_type      if case      else None,
            "judge_id":       slot.judge_id,
            "judge_name":     judge.name          if judge     else None,
            "courtroom_id":   slot.courtroom_id,
            "courtroom_name": courtroom.name      if courtroom else None,
            "start_time":     slot.start_time.isoformat(),
            "end_time":       slot.end_time.isoformat(),
            "duration_mins":  slot.duration_mins,
            "status":         "confirmed",
        })

    return {
        "id":              run_id,
        "week_start_date": hearings[0]["start_time"][:10] if hearings else None,
        "generated_at":    hearings[0]["start_time"]      if hearings else None,
        "scheduled_count": len(hearings),
        "conflict_count":  db.query(PendingQueue).count(),
        "status":          "published",
        "hearings":        hearings,
    }


@router.post("/schedules/generate", summary="Generate schedule")
async def generate_schedule_run(
    request_data: dict,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    return {
        "id":               "new-run",
        "week_start_date":  request_data.get("week_start_date", ""),
        "generated_at":     __import__('datetime').datetime.utcnow().isoformat(),
        "scheduled_count":  0,
        "unscheduled_count":0,
        "conflict_count":   0,
        "status":           "draft",
        "message":          "Use POST /api/cases/{case_id}/schedule to schedule individual cases.",
    }


@router.post("/schedules/{run_id}/publish", summary="Publish a schedule run")
def publish_schedule(
    run_id: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    return {"id": run_id, "status": "published", "message": "Schedule published successfully."}


@router.post("/schedules/generate", summary="Generate schedule (trigger AI pipeline)")
def generate_schedule(current_user=Depends(get_current_user)):
    return {"message": "Use POST /api/cases/{case_id}/schedule to schedule individual cases.",
            "status": "ok"}


def _fmt_user(u):
    return {
        "id":        u.id,
        "username":  u.username,
        "email":     u.email,
        "full_name": u.full_name,
        "role":      u.role,
        "is_active": u.is_active,
        "created_at": u.created_at.isoformat(),
    }


@router.get("/users", summary="List all users")
def list_users_alias(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.crud import get_all_users
    return [_fmt_user(u) for u in get_all_users(db)]


@router.post("/users", summary="Create a new user", status_code=201)
async def create_user_alias(
    request: Request,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.crud import get_user_by_username, create_user
    body     = await request.json()
    username = body.get("username", "").strip()
    email    = body.get("email", "").strip()
    full_name= body.get("full_name") or body.get("name", "").strip()
    password = body.get("password", "").strip()
    role     = body.get("role", "admin").strip()

    if not username or not password:
        raise HTTPException(status_code=422, detail="Username and password are required.")
    if get_user_by_username(db, username):
        raise HTTPException(status_code=409, detail="Username already exists.")

    user = create_user(db, username=username, email=email or f"{username}@court.gov.ng",
                       full_name=full_name or username, password=password, role=role)
    return _fmt_user(user)


@router.get("/users/{user_id}", summary="Get a user by ID")
def get_user_alias(
    user_id: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.crud import get_user_by_id
    user = get_user_by_id(db, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="User not found.")
    return _fmt_user(user)


@router.put("/users/{user_id}", summary="Update a user")
@router.patch("/users/{user_id}", summary="Update a user", include_in_schema=False)
async def update_user_alias(
    user_id: str,
    request: Request,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.crud import get_user_by_id
    from db.models import User
    body = await request.json()
    user = get_user_by_id(db, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="User not found.")

    if "full_name" in body: user.full_name = body["full_name"]
    if "name"      in body: user.full_name = body["name"]
    if "email"     in body: user.email     = body["email"]
    if "role"      in body: user.role      = body["role"]
    if "is_active" in body: user.is_active = body["is_active"]
    if "password"  in body and body["password"]:
        from core.security import hash_password
        user.hashed_password = hash_password(body["password"])

    db.commit()
    db.refresh(user)
    return _fmt_user(user)


@router.delete("/users/{user_id}", summary="Delete a user")
def delete_user_alias(
    user_id: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.crud import delete_user
    if not delete_user(db, user_id):
        raise HTTPException(status_code=404, detail="User not found.")
    return {"message": "User deleted successfully."}


@router.patch("/users/{user_id}/status", summary="Toggle user active status")
async def toggle_user_status(
    user_id: str,
    request: Request,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
):
    from db.crud import get_user_by_id
    body      = await request.json()
    is_active = body.get("is_active", True)
    from db.crud import update_user_status
    user = update_user_status(db, user_id, is_active)
    if not user:
        raise HTTPException(status_code=404, detail="User not found.")
    return _fmt_user(user)